File: examples/dart/wake_word_detector.dart
/// WakeWord Detector for Flutter/Dart
///
/// A cross-platform wake word detection library that works with
/// the WakeWord server for training and can run locally using ONNX.
///
/// Example usage:
/// ```dart
/// final detector = WakeWordDetector(
///   serverUrl: 'http://localhost:8000',
///   word: 'jarvis',
/// );
///
/// await detector.initialize();
/// detector.startListening(onWakeWord: () => print('Wake word detected!'));
/// ```

import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:http/http.dart' as http;
import 'package:path_provider/path_provider.dart';

// For local inference, you'll need onnxruntime_flutter package
// import 'package:onnxruntime_flutter/onnxruntime_flutter.dart';

/// Configuration for wake word detection
class WakeWordConfig {
  final String word;
  final String serverUrl;
  final double threshold;
  final int sampleRate;
  final double chunkDuration;

  const WakeWordConfig({
    required this.word,
    this.serverUrl = 'http://localhost:8000',
    this.threshold = 0.5,
    this.sampleRate = 16000,
    this.chunkDuration = 1.5,
  });
}

/// Status of a training job
class TrainingJobStatus {
  final String jobId;
  final String word;
  final String status;
  final double progress;
  final String currentStage;
  final String? errorMessage;
  final String? estimatedCompletion;
  final Map<String, dynamic> metrics;

  TrainingJobStatus({
    required this.jobId,
    required this.word,
    required this.status,
    required this.progress,
    required this.currentStage,
    this.errorMessage,
    this.estimatedCompletion,
    this.metrics = const {},
  });

  factory TrainingJobStatus.fromJson(Map<String, dynamic> json) {
    return TrainingJobStatus(
      jobId: json['job_id'] ?? '',
      word: json['word'] ?? '',
      status: json['status'] ?? '',
      progress: (json['progress'] ?? 0).toDouble(),
      currentStage: json['current_stage'] ?? '',
      errorMessage: json['error_message'],
      estimatedCompletion: json['estimated_completion'],
      metrics: json['metrics'] ?? {},
    );
  }

  bool get isComplete => status == 'completed';
  bool get isFailed => status == 'failed';
  bool get isInProgress => !isComplete && !isFailed && status != 'cancelled';
}

/// Information about an available model
class ModelInfo {
  final String word;
  final String directory;
  final Map<String, dynamic> config;
  final Map<String, dynamic> files;
  final String createdAt;

  ModelInfo({
    required this.word,
    required this.directory,
    required this.config,
    required this.files,
    required this.createdAt,
  });

  factory ModelInfo.fromJson(Map<String, dynamic> json) {
    return ModelInfo(
      word: json['word'] ?? '',
      directory: json['directory'] ?? '',
      config: json['config'] ?? {},
      files: json['files'] ?? {},
      createdAt: json['created_at'] ?? '',
    );
  }
}

/// Result of wake word prediction
class PredictionResult {
  final bool detected;
  final double confidence;
  final String word;

  PredictionResult({
    required this.detected,
    required this.confidence,
    required this.word,
  });

  factory PredictionResult.fromJson(Map<String, dynamic> json) {
    return PredictionResult(
      detected: json['detected'] ?? false,
      confidence: (json['confidence'] ?? 0).toDouble(),
      word: json['word'] ?? '',
    );
  }
}

/// Client for interacting with the WakeWord server
class WakeWordClient {
  final String serverUrl;
  final http.Client _client;

  WakeWordClient({
    this.serverUrl = 'http://localhost:8000',
    http.Client? client,
  }) : _client = client ?? http.Client();

  /// Check if the server is healthy
  Future<bool> isHealthy() async {
    try {
      final response = await _client.get(Uri.parse('$serverUrl/health'));
      return response.statusCode == 200;
    } catch (e) {
      return false;
    }
  }

  /// List all available models
  Future<List<ModelInfo>> listModels() async {
    final response = await _client.get(Uri.parse('$serverUrl/models'));

    if (response.statusCode != 200) {
      throw Exception('Failed to list models: ${response.body}');
    }

    final data = jsonDecode(response.body);
    final models =
        (data['models'] as List).map((m) => ModelInfo.fromJson(m)).toList();

    return models;
  }

  /// Check if a model exists for a word
  Future<bool> modelExists(String word) async {
    try {
      final response = await _client.get(
        Uri.parse('$serverUrl/models/$word/config'),
      );
      return response.statusCode == 200;
    } catch (e) {
      return false;
    }
  }

  /// Request training for a new wake word model
  Future<TrainingJobStatus> requestTraining({
    required String word,
    int sampleCount = 500,
    int epochs = 50,
    int batchSize = 32,
  }) async {
    final response = await _client.post(
      Uri.parse('$serverUrl/train'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'word': word,
        'sample_count': sampleCount,
        'epochs': epochs,
        'batch_size': batchSize,
      }),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to request training: ${response.body}');
    }

    return TrainingJobStatus.fromJson(jsonDecode(response.body));
  }

  /// Get the status of a training job
  Future<TrainingJobStatus> getJobStatus(String jobId) async {
    final response = await _client.get(Uri.parse('$serverUrl/jobs/$jobId'));

    if (response.statusCode != 200) {
      throw Exception('Failed to get job status: ${response.body}');
    }

    return TrainingJobStatus.fromJson(jsonDecode(response.body));
  }

  /// Wait for training to complete
  Future<TrainingJobStatus> waitForTraining(
    String jobId, {
    Duration pollInterval = const Duration(seconds: 10),
    Duration timeout = const Duration(minutes: 30),
    void Function(TrainingJobStatus)? onProgress,
  }) async {
    final deadline = DateTime.now().add(timeout);

    while (DateTime.now().isBefore(deadline)) {
      final status = await getJobStatus(jobId);

      if (onProgress != null) {
        onProgress(status);
      }

      if (status.isComplete) {
        return status;
      }

      if (status.isFailed) {
        throw Exception('Training failed: ${status.errorMessage}');
      }

      await Future.delayed(pollInterval);
    }

    throw TimeoutException('Training timed out');
  }

  /// Download a model file
  Future<Uint8List> downloadModel(String word, {String format = 'onnx'}) async {
    final response = await _client.get(
      Uri.parse('$serverUrl/models/$word?format=$format'),
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to download model: ${response.body}');
    }

    return response.bodyBytes;
  }

  /// Download model to a file
  Future<File> downloadModelToFile(
    String word, {
    String format = 'onnx',
    String? directory,
  }) async {
    final bytes = await downloadModel(word, format: format);

    final dir = directory ?? (await getApplicationDocumentsDirectory()).path;
    final file = File('$dir/${word}_model.$format');
    await file.writeAsBytes(bytes);

    return file;
  }

  /// Ensure a model exists, training if necessary
  Future<void> ensureModel(
    String word, {
    int sampleCount = 500,
    int epochs = 50,
    void Function(TrainingJobStatus)? onProgress,
  }) async {
    if (await modelExists(word)) {
      return;
    }

    final job = await requestTraining(
      word: word,
      sampleCount: sampleCount,
      epochs: epochs,
    );

    if (job.isComplete) {
      return;
    }

    await waitForTraining(job.jobId, onProgress: onProgress);
  }

  /// Predict if audio contains the wake word (server-side)
  Future<PredictionResult> predict(String word, Uint8List audioBytes) async {
    final audioBase64 = base64Encode(audioBytes);

    final response = await _client.post(
      Uri.parse('$serverUrl/predict/$word'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'audio_base64': audioBase64}),
    );

    if (response.statusCode != 200) {
      throw Exception('Prediction failed: ${response.body}');
    }

    return PredictionResult.fromJson(jsonDecode(response.body));
  }

  void dispose() {
    _client.close();
  }
}

/// Wake word detector with microphone input
///
/// This is a basic structure - actual implementation requires
/// platform-specific audio recording (e.g., flutter_sound, record)
abstract class WakeWordDetector {
  final WakeWordConfig config;
  final WakeWordClient client;

  bool _isListening = false;

  WakeWordDetector({required this.config, WakeWordClient? client})
    : client = client ?? WakeWordClient(serverUrl: config.serverUrl);

  bool get isListening => _isListening;

  /// Initialize the detector (download model, set up audio, etc.)
  Future<void> initialize();

  /// Start listening for the wake word
  Future<void> startListening({
    required void Function() onWakeWord,
    void Function(double confidence)? onPrediction,
    void Function(Object error)? onError,
  });

  /// Stop listening
  Future<void> stopListening();

  /// Dispose resources
  void dispose() {
    stopListening();
    client.dispose();
  }
}

/// Server-based wake word detector
///
/// Sends audio to server for prediction.
/// Simpler to implement but requires network connectivity.
class ServerWakeWordDetector extends WakeWordDetector {
  StreamSubscription? _audioSubscription;

  ServerWakeWordDetector({
    required WakeWordConfig config,
    WakeWordClient? client,
  }) : super(config: config, client: client);

  @override
  Future<void> initialize() async {
    // Ensure model exists on server
    await client.ensureModel(
      config.word,
      onProgress: (status) {
        print(
          'Training progress: ${status.progress}% - ${status.currentStage}',
        );
      },
    );
  }

  @override
  Future<void> startListening({
    required void Function() onWakeWord,
    void Function(double confidence)? onPrediction,
    void Function(Object error)? onError,
  }) async {
    if (_isListening) return;
    _isListening = true;

    // TODO: Implement actual audio recording using platform-specific packages
    // Example with flutter_sound or record package:
    //
    // final recorder = FlutterSoundRecorder();
    // await recorder.openRecorder();
    //
    // Timer.periodic(Duration(milliseconds: (config.chunkDuration * 1000).toInt()), (timer) async {
    //   if (!_isListening) {
    //     timer.cancel();
    //     return;
    //   }
    //
    //   try {
    //     final audioBytes = await recorder.getAudioChunk();
    //     final result = await client.predict(config.word, audioBytes);
    //
    //     if (onPrediction != null) {
    //       onPrediction(result.confidence);
    //     }
    //
    //     if (result.detected && result.confidence >= config.threshold) {
    //       onWakeWord();
    //     }
    //   } catch (e) {
    //     if (onError != null) {
    //       onError(e);
    //     }
    //   }
    // });

    print('Listening for wake word "${config.word}"...');
    print(
      'Note: Audio recording not implemented - use flutter_sound or record package',
    );
  }

  @override
  Future<void> stopListening() async {
    _isListening = false;
    await _audioSubscription?.cancel();
    _audioSubscription = null;
  }
}

/// Local wake word detector using ONNX Runtime
///
/// Runs inference locally without network connectivity.
/// Requires onnxruntime_flutter package.
class LocalWakeWordDetector extends WakeWordDetector {
  File? _modelFile;
  // OrtSession? _session;  // Uncomment when using onnxruntime_flutter

  LocalWakeWordDetector({
    required WakeWordConfig config,
    WakeWordClient? client,
  }) : super(config: config, client: client);

  @override
  Future<void> initialize() async {
    // Download model if needed
    _modelFile = await client.downloadModelToFile(config.word, format: 'onnx');

    // Initialize ONNX Runtime
    // Uncomment when using onnxruntime_flutter:
    // await OrtEnv.instance.init();
    // _session = OrtSession.fromFile(_modelFile!.path);

    print('Local model loaded from: ${_modelFile!.path}');
  }

  @override
  Future<void> startListening({
    required void Function() onWakeWord,
    void Function(double confidence)? onPrediction,
    void Function(Object error)? onError,
  }) async {
    if (_isListening) return;
    _isListening = true;

    // TODO: Implement local inference with ONNX Runtime
    // Example:
    //
    // Timer.periodic(Duration(milliseconds: (config.chunkDuration * 1000).toInt()), (timer) async {
    //   if (!_isListening) {
    //     timer.cancel();
    //     return;
    //   }
    //
    //   try {
    //     final audioData = await getAudioChunk();
    //     final mfcc = extractMFCC(audioData);
    //
    //     final inputs = {'input': OrtValueTensor.fromList(mfcc, [1, 40, 150])};
    //     final outputs = await _session!.run(inputs);
    //     final confidence = outputs['output']!.value[0];
    //
    //     if (onPrediction != null) {
    //       onPrediction(confidence);
    //     }
    //
    //     if (confidence >= config.threshold) {
    //       onWakeWord();
    //     }
    //   } catch (e) {
    //     if (onError != null) {
    //       onError(e);
    //     }
    //   }
    // });

    print('Listening for wake word "${config.word}" (local mode)...');
    print(
      'Note: ONNX inference not implemented - use onnxruntime_flutter package',
    );
  }

  @override
  Future<void> stopListening() async {
    _isListening = false;
  }

  @override
  void dispose() {
    // _session?.release();
    super.dispose();
  }
}

// ============================================================================
// Example Usage
// ============================================================================

void main() async {
  // Example: Using server-based detection
  final config = WakeWordConfig(
    word: 'jarvis',
    serverUrl: 'http://localhost:8000',
    threshold: 0.5,
  );

  final detector = ServerWakeWordDetector(config: config);

  try {
    print('Initializing wake word detector...');
    await detector.initialize();

    print('Starting to listen...');
    await detector.startListening(
      onWakeWord: () {
        print('🎤 Wake word detected!');
      },
      onPrediction: (confidence) {
        if (confidence > 0.3) {
          print('Confidence: ${(confidence * 100).toStringAsFixed(1)}%');
        }
      },
      onError: (error) {
        print('Error: $error');
      },
    );

    // Keep running for demo
    await Future.delayed(Duration(minutes: 5));
  } finally {
    detector.dispose();
  }
}

File: examples/android/WakeWordDetector.kt
package com.wakeword.detector

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import android.util.Base64
import android.util.Log
import androidx.core.app.ActivityCompat
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * WakeWord Detector for Android
 *
 * Detects custom wake words using either:
 * 1. Server-based inference (sends audio to WakeWord server)
 * 2. Local inference using ONNX Runtime (offline capable)
 *
 * Usage:
 * ```kotlin
 * val detector = WakeWordDetector(context, "http://server:8000", "jarvis")
 * detector.initialize()
 * detector.startListening { confidence ->
 *     if (confidence > 0.5) {
 *         Log.d("WakeWord", "Detected!")
 *     }
 * }
 * ```
 */

// ============================================================================
// Data Classes
// ============================================================================

data class WakeWordConfig(
    val word: String,
    val serverUrl: String = "http://localhost:8000",
    val threshold: Float = 0.5f,
    val sampleRate: Int = 16000,
    val chunkDurationMs: Int = 1500,
)

data class TrainingStatus(
    val jobId: String,
    val word: String,
    val status: String,
    val progress: Float,
    val currentStage: String,
    val errorMessage: String? = null,
) {
    val isComplete: Boolean get() = status == "completed"
    val isFailed: Boolean get() = status == "failed"
    val isInProgress: Boolean get() = !isComplete && !isFailed && status != "cancelled"
}

data class PredictionResult(
    val detected: Boolean,
    val confidence: Float,
    val word: String,
)

sealed class DetectorState {
    object Uninitialized : DetectorState()
    object Initializing : DetectorState()
    object Ready : DetectorState()
    object Listening : DetectorState()
    data class Error(val message: String) : DetectorState()
}

// ============================================================================
// WakeWord Client
// ============================================================================

class WakeWordClient(private val serverUrl: String) {
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
        .readTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
        .build()

    companion object {
        private const val TAG = "WakeWordClient"
    }

    /**
     * Check if server is healthy
     */
    suspend fun isHealthy(): Boolean = withContext(Dispatchers.IO) {
        try {
            val request = Request.Builder()
                .url("$serverUrl/health")
                .get()
                .build()
            
            val response = client.newCall(request).execute()
            response.isSuccessful
        } catch (e: Exception) {
            Log.e(TAG, "Health check failed: ${e.message}")
            false
        }
    }

    /**
     * Check if model exists for a word
     */
    suspend fun modelExists(word: String): Boolean = withContext(Dispatchers.IO) {
        try {
            val request = Request.Builder()
                .url("$serverUrl/models/$word/config")
                .get()
                .build()
            
            val response = client.newCall(request).execute()
            response.isSuccessful
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Request training for a word
     */
    suspend fun requestTraining(
        word: String,
        sampleCount: Int = 500,
        epochs: Int = 50,
    ): TrainingStatus = withContext(Dispatchers.IO) {
        val json = JSONObject().apply {
            put("word", word)
            put("sample_count", sampleCount)
            put("epochs", epochs)
        }

        val request = Request.Builder()
            .url("$serverUrl/train")
            .post(json.toString().toRequestBody("application/json".toMediaType()))
            .build()

        val response = client.newCall(request).execute()
        if (!response.isSuccessful) {
            throw IOException("Training request failed: ${response.body?.string()}")
        }

        val body = JSONObject(response.body?.string() ?: "{}")
        TrainingStatus(
            jobId = body.optString("job_id", ""),
            word = body.optString("word", word),
            status = body.optString("status", "pending"),
            progress = body.optDouble("progress", 0.0).toFloat(),
            currentStage = body.optString("current_stage", ""),
        )
    }

    /**
     * Get training job status
     */
    suspend fun getJobStatus(jobId: String): TrainingStatus = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("$serverUrl/jobs/$jobId")
            .get()
            .build()

        val response = client.newCall(request).execute()
        if (!response.isSuccessful) {
            throw IOException("Failed to get job status: ${response.body?.string()}")
        }

        val body = JSONObject(response.body?.string() ?: "{}")
        TrainingStatus(
            jobId = body.optString("job_id", jobId),
            word = body.optString("word", ""),
            status = body.optString("status", ""),
            progress = body.optDouble("progress", 0.0).toFloat(),
            currentStage = body.optString("current_stage", ""),
            errorMessage = body.optString("error_message", null),
        )
    }

    /**
     * Wait for training to complete
     */
    suspend fun waitForTraining(
        jobId: String,
        pollIntervalMs: Long = 10000,
        onProgress: ((TrainingStatus) -> Unit)? = null,
    ): TrainingStatus {
        while (true) {
            val status = getJobStatus(jobId)
            onProgress?.invoke(status)

            when {
                status.isComplete -> return status
                status.isFailed -> throw IOException("Training failed: ${status.errorMessage}")
            }

            delay(pollIntervalMs)
        }
    }

    /**
     * Download model bytes
     */
    suspend fun downloadModel(word: String, format: String = "onnx"): ByteArray = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("$serverUrl/models/$word?format=$format")
            .get()
            .build()

        val response = client.newCall(request).execute()
        if (!response.isSuccessful) {
            throw IOException("Failed to download model: ${response.body?.string()}")
        }

        response.body?.bytes() ?: throw IOException("Empty response")
    }

    /**
     * Predict wake word from audio
     */
    suspend fun predict(word: String, audioBytes: ByteArray): PredictionResult = withContext(Dispatchers.IO) {
        val audioBase64 = Base64.encodeToString(audioBytes, Base64.NO_WRAP)
        
        val json = JSONObject().apply {
            put("audio_base64", audioBase64)
        }

        val request = Request.Builder()
            .url("$serverUrl/predict/$word")
            .post(json.toString().toRequestBody("application/json".toMediaType()))
            .build()

        val response = client.newCall(request).execute()
        if (!response.isSuccessful) {
            throw IOException("Prediction failed: ${response.body?.string()}")
        }

        val body = JSONObject(response.body?.string() ?: "{}")
        PredictionResult(
            detected = body.optBoolean("detected", false),
            confidence = body.optDouble("confidence", 0.0).toFloat(),
            word = body.optString("word", word),
        )
    }

    /**
     * Ensure model exists, training if necessary
     */
    suspend fun ensureModel(
        word: String,
        sampleCount: Int = 500,
        epochs: Int = 50,
        onProgress: ((TrainingStatus) -> Unit)? = null,
    ) {
        if (modelExists(word)) {
            Log.d(TAG, "Model for '$word' already exists")
            return
        }

        Log.d(TAG, "Requesting training for '$word'...")
        val job = requestTraining(word, sampleCount, epochs)
        
        if (!job.isComplete) {
            waitForTraining(job.jobId, onProgress = onProgress)
        }
    }
}

// ============================================================================
// Audio Recorder
// ============================================================================

class AudioRecorder(
    private val sampleRate: Int = 16000,
    private val chunkDurationMs: Int = 1500,
) {
    private var audioRecord: AudioRecord? = null
    private var isRecording = false
    private var recordingJob: Job? = null

    val bufferSize: Int
        get() = AudioRecord.getMinBufferSize(
            sampleRate,
            AudioFormat.CHANNEL_IN_MONO,
            AudioFormat.ENCODING_PCM_16BIT
        ).coerceAtLeast(sampleRate * chunkDurationMs / 1000 * 2)

    fun start(context: Context, onAudioChunk: (ByteArray) -> Unit): Boolean {
        if (ActivityCompat.checkSelfPermission(
                context, Manifest.permission.RECORD_AUDIO
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            return false
        }

        val chunkSamples = sampleRate * chunkDurationMs / 1000
        val chunkBytes = chunkSamples * 2  // 16-bit = 2 bytes

        audioRecord = AudioRecord(
            MediaRecorder.AudioSource.MIC,
            sampleRate,
            AudioFormat.CHANNEL_IN_MONO,
            AudioFormat.ENCODING_PCM_16BIT,
            bufferSize
        )

        if (audioRecord?.state != AudioRecord.STATE_INITIALIZED) {
            return false
        }

        isRecording = true
        audioRecord?.startRecording()

        recordingJob = CoroutineScope(Dispatchers.IO).launch {
            val buffer = ShortArray(chunkSamples)
            
            while (isRecording) {
                val read = audioRecord?.read(buffer, 0, chunkSamples) ?: 0
                if (read > 0) {
                    // Convert to WAV bytes
                    val wavBytes = pcmToWav(buffer, read, sampleRate)
                    onAudioChunk(wavBytes)
                }
            }
        }

        return true
    }

    fun stop() {
        isRecording = false
        recordingJob?.cancel()
        audioRecord?.stop()
        audioRecord?.release()
        audioRecord = null
    }

    private fun pcmToWav(pcmData: ShortArray, numSamples: Int, sampleRate: Int): ByteArray {
        val byteRate = sampleRate * 2  // 16-bit mono
        val dataSize = numSamples * 2
        val totalSize = 44 + dataSize

        val buffer = ByteBuffer.allocate(totalSize).order(ByteOrder.LITTLE_ENDIAN)

        // RIFF header
        buffer.put("RIFF".toByteArray())
        buffer.putInt(totalSize - 8)
        buffer.put("WAVE".toByteArray())

        // fmt chunk
        buffer.put("fmt ".toByteArray())
        buffer.putInt(16)  // chunk size
        buffer.putShort(1)  // PCM format
        buffer.putShort(1)  // mono
        buffer.putInt(sampleRate)
        buffer.putInt(byteRate)
        buffer.putShort(2)  // block align
        buffer.putShort(16)  // bits per sample

        // data chunk
        buffer.put("data".toByteArray())
        buffer.putInt(dataSize)
        
        for (i in 0 until numSamples) {
            buffer.putShort(pcmData[i])
        }

        return buffer.array()
    }
}

// ============================================================================
// Wake Word Detector
// ============================================================================

class WakeWordDetector(
    private val context: Context,
    private val config: WakeWordConfig,
) {
    private val client = WakeWordClient(config.serverUrl)
    private val audioRecorder = AudioRecorder(config.sampleRate, config.chunkDurationMs)
    
    private val _state = MutableStateFlow<DetectorState>(DetectorState.Uninitialized)
    val state: StateFlow<DetectorState> = _state

    private var detectionJob: Job? = null

    companion object {
        private const val TAG = "WakeWordDetector"
    }

    /**
     * Initialize the detector
     */
    suspend fun initialize(
        onProgress: ((TrainingStatus) -> Unit)? = null,
    ) {
        _state.value = DetectorState.Initializing
        
        try {
            // Check server health
            if (!client.isHealthy()) {
                throw IOException("Server is not reachable")
            }

            // Ensure model exists
            client.ensureModel(
                config.word,
                onProgress = onProgress,
            )

            _state.value = DetectorState.Ready
            Log.d(TAG, "Detector initialized for '${config.word}'")
        } catch (e: Exception) {
            Log.e(TAG, "Initialization failed: ${e.message}")
            _state.value = DetectorState.Error(e.message ?: "Unknown error")
            throw e
        }
    }

    /**
     * Start listening for wake word
     */
    fun startListening(
        onWakeWord: (Float) -> Unit,
        onError: ((Exception) -> Unit)? = null,
    ): Boolean {
        if (_state.value != DetectorState.Ready) {
            Log.w(TAG, "Detector not ready")
            return false
        }

        val started = audioRecorder.start(context) { audioBytes ->
            detectionJob = CoroutineScope(Dispatchers.IO).launch {
                try {
                    val result = client.predict(config.word, audioBytes)
                    
                    if (result.detected && result.confidence >= config.threshold) {
                        Log.d(TAG, "Wake word detected! Confidence: ${result.confidence}")
                        withContext(Dispatchers.Main) {
                            onWakeWord(result.confidence)
                        }
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "Prediction error: ${e.message}")
                    withContext(Dispatchers.Main) {
                        onError?.invoke(e)
                    }
                }
            }
        }

        if (started) {
            _state.value = DetectorState.Listening
            Log.d(TAG, "Started listening for '${config.word}'")
        }

        return started
    }

    /**
     * Stop listening
     */
    fun stopListening() {
        audioRecorder.stop()
        detectionJob?.cancel()
        
        if (_state.value == DetectorState.Listening) {
            _state.value = DetectorState.Ready
        }
        
        Log.d(TAG, "Stopped listening")
    }

    /**
     * Release resources
     */
    fun release() {
        stopListening()
        _state.value = DetectorState.Uninitialized
    }
}

// ============================================================================
// Local Detector (ONNX Runtime)
// ============================================================================

/**
 * Local wake word detector using ONNX Runtime
 * 
 * For local inference, add these dependencies:
 * implementation 'com.microsoft.onnxruntime:onnxruntime-android:1.16.0'
 */
class LocalWakeWordDetector(
    private val context: Context,
    private val config: WakeWordConfig,
) {
    private var modelBytes: ByteArray? = null
    // private var ortSession: OrtSession? = null  // Uncomment with ONNX Runtime
    
    private val audioRecorder = AudioRecorder(config.sampleRate, config.chunkDurationMs)
    private val client = WakeWordClient(config.serverUrl)
    
    private val _state = MutableStateFlow<DetectorState>(DetectorState.Uninitialized)
    val state: StateFlow<DetectorState> = _state

    companion object {
        private const val TAG = "LocalWakeWordDetector"
    }

    /**
     * Initialize with local model
     */
    suspend fun initialize() {
        _state.value = DetectorState.Initializing

        try {
            // Try to load cached model
            val modelFile = File(context.filesDir, "${config.word}_model.onnx")
            
            if (modelFile.exists()) {
                modelBytes = modelFile.readBytes()
            } else {
                // Download from server
                Log.d(TAG, "Downloading model...")
                modelBytes = client.downloadModel(config.word, "onnx")
                modelFile.writeBytes(modelBytes!!)
            }

            // Initialize ONNX Runtime
            // Uncomment when using ONNX Runtime:
            // val ortEnv = OrtEnvironment.getEnvironment()
            // ortSession = ortEnv.createSession(modelBytes)

            _state.value = DetectorState.Ready
            Log.d(TAG, "Local detector initialized")
        } catch (e: Exception) {
            _state.value = DetectorState.Error(e.message ?: "Unknown error")
            throw e
        }
    }

    /**
     * Start listening (local inference)
     */
    fun startListening(
        onWakeWord: (Float) -> Unit,
        onError: ((Exception) -> Unit)? = null,
    ): Boolean {
        if (_state.value != DetectorState.Ready) {
            return false
        }

        val started = audioRecorder.start(context) { audioBytes ->
            CoroutineScope(Dispatchers.Default).launch {
                try {
                    val confidence = runLocalInference(audioBytes)
                    
                    if (confidence >= config.threshold) {
                        Log.d(TAG, "Wake word detected locally! Confidence: $confidence")
                        withContext(Dispatchers.Main) {
                            onWakeWord(confidence)
                        }
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "Local inference error: ${e.message}")
                    withContext(Dispatchers.Main) {
                        onError?.invoke(e)
                    }
                }
            }
        }

        if (started) {
            _state.value = DetectorState.Listening
        }

        return started
    }

    private fun runLocalInference(audioBytes: ByteArray): Float {
        // TODO: Implement MFCC extraction and ONNX inference
        // 
        // val mfcc = extractMFCC(audioBytes)
        // val inputTensor = OnnxTensor.createTensor(ortEnv, mfcc)
        // val results = ortSession?.run(mapOf("input" to inputTensor))
        // return results?.get(0)?.value as Float
        
        return 0f  // Placeholder
    }

    fun stopListening() {
        audioRecorder.stop()
        if (_state.value == DetectorState.Listening) {
            _state.value = DetectorState.Ready
        }
    }

    fun release() {
        stopListening()
        // ortSession?.close()
        _state.value = DetectorState.Uninitialized
    }
}
